### PR TITLE
Deprecate legacy cookie behaviour behind `LEGACY_COOKIES` profile

### LINKED ISSUE #16181: Deprecate legacy cookie behaviour behind profile
### Description

Since implementing [KEYCLOAK-12125](https://issues.redhat.com/browse/KEYCLOAK-12125), Keycloak adds two cookies with the same value but different `SameSite` settings when the `SameSite=None` is set to none. 

The was due to old browsers not support `SameSite=None`, see https://www.chromium.org/updates/same-site/incompatible-clients/ for details and how to detect them.

Since then, browsers supporting this are wide-spread: https://caniuse.com/?search=SameSite%3DNone (92+%). 

Adding these extra cookies adds to the HTTP headers sent in a response and request, and a default setup with NGINX which limits responses with 4k headers needs to be tweaked to make it work. 

Therefore, I suggest to remove the `*_LEGACY` cookies. Additionally, we should unify our cookie behaviour so that:
- All cookies are `Secure` by default.
- All cookies are `SameSite=Strict`, unless explicitly specified otherwise.

### Motivation

* Simplify Keycloak's code structure, and support ever-green browers.
* Simplify setup of Keycloak behind a loadbalancer

### PR DESCRIPTION
**Pull request by jonkoops**
Closes #16181

### CODE REVIEW

**Comment by ahus1** [comment]
Happy to help if you want. I suppose there will be some failing tests as some Quarkus tests list all options.

**Comment by jonkoops** [comment]
*Warning, rant below, nothing personal, just need to vent. Consider this off-topic.*

> I suppose there will be some failing tests as some Quarkus tests list all options in the help.

This remains one of the most problematic things about contributing to Keycloak for me. I actively don't want to touch anything related to `Profile`, as I have to deal with this incredibly frustrating part of the test suite. I'd really appreciate it if we didn't have to do this every time a Profile changes, it's a terrible DX.

It's also incredibly manual. Every time this changes (and this is merge conflict prone) I have to regenerate these files and manually rename them. Why can't this be done by [ApprovalTests](https://github.com/approvals/ApprovalTests.Java) itself with some kind of flag like `-PupdateApprovals` or something?

It's also required to generate the files for both Unix and Windows systems, the latter I don't have. So I end up having to copy the contents from the Unix version and fix a bunch of stuff like line endings manually.

I understand the need to test this functionality, but can we do something here to improve the DX, like:

- Adding the aforementioned profile to automatically update the `.approved` snapshots
- Generating the Windows versions for these files, even from a Unix system

**Comment by jonkoops** [comment]
I am surprised, the tests are passing, but I have not updated the Windows snapshots. Are these not being tested?

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (jonkoops) [inline docs/documentation/release_notes/topics/24_0_0.adoc:1]: @andymunro could I ask you to vet the changes to the release notes for me?

**Comment by jonkoops** [comment]
@vmuzikar You had some concerns on the previous PR (https://github.com/keycloak/keycloak/pull/16747#pullrequestreview-1284992026), let's try to resolve those here.

> I'm afraid removing legacy cookies breaks HTTP (non-SSL) requests, at least to the admin console. That is because we have to force `Secure` anytime we use `SameSite=None`:

I think we need to stop supporting this use-case, folks should not be running Keycloak on HTTP unless they are behind a proxy that is serving over HTTPS, in which case this is also not a problem, as this would be considered a secure context.

I also don't expect this to be a problem for local development, as long as folks are using a [secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#when_is_a_context_considered_secure), such as `http://localhost` and `http://*.localhost` URLs (e.g. `http://dev.whatever.localhost/`).

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testDefaultToHelp.unix.appoved.txt
Comment (andymunro) [inline quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testDefaultToHelp.unix.appoved.txt:49]:
```suggestion
  See the documentation for more details before deploying in
```

Comment (andymunro) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testDefaultToHelp.unix.appoved.txt:49]: Our standard tone is to omit "Please."  Also, we use "see" instead of "take a look"

**Review by andymunro — CHANGES_REQUESTED** [review]
Looks good. Just one suggestion @jonkoops

**Comment by jonkoops** [comment]
Thanks @andymunro, I've incorporated your feedback. Could you review the changes?

**Review by pedroigor — COMMENTED** [review]
Can we make the changes not related to the original issue as a set of follow-ups?

**Comment by pedroigor** [comment]
@mposolda Wdyt? It sounds like a good plan until finally deprecated it.

**Comment by jonkoops** [comment]
@andymunro I split out your feedback into a separate PR to clean this one up a little, could you review it for me? https://github.com/keycloak/keycloak/pull/26173

**Comment by jonkoops** [comment]
> Can we make the changes not related to the original issue as a set of follow-ups?

Done, can you take another look at the changes?

**Review by andymunro — APPROVED** [review]
Looks good, @jonkoops

**Comment by jonkoops** [comment]
Thanks Andrew!

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (hmlnarik) [inline docs/documentation/release_notes/topics/24_0_0.adoc:8]: `start-dev` is for developers while release notes target non-dev users. Perhaps better to use `start` command here.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/UserStorageTest.java
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/UserStorageTest.java:107]: Since the legacy cookies are going to be removed, could you instead update the `testLoginSuccessWithSpecialCharacter` not to check legacy cookie?

**Review by mposolda — DISMISSED** [review]
@jonkoops Looks good, Thanks!

**Review by edewit — DISMISSED** [review]

**Review by hmlnarik — COMMENTED** [review]
The direction looks good, please see few comments inline.
Could you please also remove references to legacy cookies in `KcOidcBrokerLogoutTest`, `SAMLServletAdapterTest`, `AuthenticationSessionFailoverClusterTest`

**Review by ahus1 — CHANGES_REQUESTED** [review]
When you run Keycloak, it will print 

```
 Keycloak 999.0.0-SNAPSHOT on JVM (powered by Quarkus 3.2.9.Final) started in 12.006s. Listening on: http://0.0.0.0:8080
``` 

And after the change you won't be able to click on that URL and log in. Instead you'll see an error "This page isn’t working" and "ERR_TOO_MANY_REDIRECTS" after entering the username and password.

This is the same if someone opens the URL for whatever reason without `https://`.

This is a UX that IMHO needs improvement, as it is difficult to understand for the user what's going wrong here, especially for users new to Keycloak.

* Before showing the login screen (or after submitting the form the lastest), check if cookies are working as expected. If they don't, show a descriptive error message instead of the login screen 
* Offer a CLI option to allow for non-secure cookies (for example `--http-cookies-secure=false`)

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/UserStorageTest.java
Comment (jonkoops) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/UserStorageTest.java:107]: I'd rather keep it like this and remove both the profile and test code at a later point in time. Even thought we're deprecating this profile, we should still test this functionality while the profile exists.

**Comment by jonkoops** [comment]
> This is a UX that IMHO needs improvement, as it is difficult to understand for the user what's going wrong here, especially for users new to Keycloak.

@ahus1 I agree with you that this is problematic, we should not be presenting URLs to the user if they cannot be used, or at least not without also accompanying it with a proper disclaimer (both at runtime and in the documentation). Note that everything works fine on a [secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#when_is_a_context_considered_secure), e.g. `http://locahost:8080` is considered 'secure'.

I'll take a look to see if we can improve this under this PR before merging it in, at the very least this should be reflected in the release notes I think.

> Before showing the login screen (or after submitting the form the lastest), check if cookies are working as expected. If they don't, show a descriptive error message instead of the login screen

I think we can use [`window.isSecureContext`](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#feature_detection) to detect this with JavaScript and show a banner on the login page. Perhaps there is also a way to detect this on the server-side, but this seems the easiest solution. Should we take this on separately of this PR?


> Offer a CLI option to allow for non-secure cookies (for example `--http-cookies-secure=false`)

On this we disagree. The reason for this is that various APIs and functionalities of the browser are simply [not available](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts/features_restricted_to_secure_contexts) in an insecure context, requiring us to handle even more permutations in an already complex stack. I don't think it's considered reasonable for anyone to run Keycloak without also having it operate in a secure environment, even in development there are options (e.g. `http://dev.whatever.localhost/`).

I do agree that this should be properly documented, and this PR leaves that lacking. I will put in the work to get that resolved.

**Comment by jonkoops** [comment]
I've converted this PR to a draft, as I am feeling this is going to be a larger effort around making cookies in Keycloak behave in a unified and secure manner in general.

You've all probably seen the various warnings around cookies in browser consoles, and there are a lot of changes coming up this year to browsers and their cookie policies this year. We should be ahead of the curve here, so I am going to take this an opportunity to do some deeper research and try to get us in a good state in this regard.

**Comment by ahus1** [comment]
@jonkoops - thank you for your reply. 

> The reason for this is that various APIs and functionalities of the browser are simply not available in an insecure context, requiring us to handle even more permutations in an already complex stack.

If this is the case, I agree to not have such a CLI option. The litmus test is how we can put this into release notes and into a warning message for users so they understand and agree. 

> I think we can use [window.isSecureContext](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#feature_detection) to detect this with JavaScript and show a banner on the login page.

I would go as far as disabling the login form in such a case. If we get the UX right with JavaScript only, we could proceed.

> You've all probably seen the various warnings around cookies in browser consoles, and there are a lot of changes coming up this year to browsers and their cookie policies this year. We should be ahead of the curve here, so I am going to take this an opportunity to do some deeper research and try to get us in a good state in this regard.

I totally agree on that, and I'm happy to help where I can around this.

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (jonkoops) [reply docs/documentation/release_notes/topics/24_0_0.adoc]: Should be resolved now :+1:

**Comment by pedroigor** [comment]
In regards to the redirect issue when using insecure connections, IMO it is time to force people to enable TLS at production. However, we need to look at some of the issues we had in the past around hostname because when we tried to do that in the past we had quite a few issues/complains about our assumption not being applicable for some use cases. I don't remember what these use cases are but I think it is worth to look at them. @vmuzikar wdyt?

As per the documentation, we should highlight the implications of not enabling TLS and make clearer what to expect when not enabling it in production using a valid domain.

I like @ahus1 suggestion to improve UX by recognizing whether the server will work with insecure connections. We can that at the server-side, I think. For instance, checking both HTTPS and Hostname settings we should be able to print a warning in the logs. At runtime, we could show a proper error page. Works for me if we do this is a follow-up, but a more clear documentation about the topic is a must.

If we want to improve the JS adapter, that would be nice too.

**Comment by jonkoops** [comment]
> If we want to improve the JS adapter, that would be nice too.

Yeah, this has been on hold for quite some time, also due to the Privacy Sandbox changes coming up, which will essentially cripple some parts of the JS adapter. You can read more about that effort under #16834

**Comment by jonkoops** [comment]
I have updated the release notes with further information on how I envision the rest of the changes in this PR to look, feel free to comment and provide reviews on this text.

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (pedroigor) [inline docs/documentation/release_notes/topics/24_0_0.adoc:9]: Is this title accurate or it is more about our future plans? We don't necessarily need to force `strict` but also consider making the attribute configurable, as per discussions we had in the past.

**Comment by pedroigor** [comment]
@jonkoops Do we also want to update our TLS or production guides?

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (jonkoops) [reply docs/documentation/release_notes/topics/24_0_0.adoc:9]: The documentation updates here reflect what I intend the implementation to be under this PR (yet to be implemented). 

> We don't necessarily need to force `strict` but also consider making the attribute configurable, as per discussions we had in the past.

It would still be 'overwritable' in code for specific use-cases, but not globally configurable if that is what you mean here. Is there a link to these previous discussions somewhere that can be referenced?

Future changes in Chrome will move all these to `Strict` by default, as part of the 3rd-party cookie phase-out. So this is just about getting ahead of these changes by forcing it as the default.

**Comment by jonkoops** [comment]
> Do we also want to update our TLS or production guides?

Yes, absolutely. We'll have to comb the documentation and make sure we're enforcing a secure context everywhere.

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (pedroigor) [reply docs/documentation/release_notes/topics/24_0_0.adoc:9]: > Future changes in Chrome will move all these to Strict by default, as part of the 3rd-party cookie phase-out. So this is just about getting ahead of these changes by forcing it as the default.

Yeah, but only if we don't set the attribute, right? My point is that I'm not sure if we should do the `Strict` now because, AFAIK, it is perfectly fine to go with `secure` and `none`. And TBH, I think that should be the default as a balance between security and usability.

I'll need to go over the closed/open issues to filter discussions ...

**Comment by stianst** [comment]
> When you run Keycloak, it will print
> 
> ```
>  Keycloak 999.0.0-SNAPSHOT on JVM (powered by Quarkus 3.2.9.Final) started in 12.006s. Listening on: http://0.0.0.0:8080
> ```
> 
> And after the change you won't be able to click on that URL and log in. Instead you'll see an error "This page isn’t working" and "ERR_TOO_MANY_REDIRECTS" after entering the username and password.
> 
> This is the same if someone opens the URL for whatever reason without `https://`.
> 
> This is a UX that IMHO needs improvement, as it is difficult to understand for the user what's going wrong here, especially for users new to Keycloak.
> 
> * Before showing the login screen (or after submitting the form the lastest), check if cookies are working as expected. If they don't, show a descriptive error message instead of the login screen
> * Offer a CLI option to allow for non-secure cookies (for example `--http-cookies-secure=false`)

Thinking we shouldn't output `http://0.0.0.0:8080` that's not really an actual URL. Perhaps we should drop this from the Keycloak started, and have some output from the `hostname` provider instead? @pedroigor @vmuzikar ?

**Comment by jonkoops** [comment]
> Thinking we shouldn't output `http://0.0.0.0:8080` that's not really an actual URL. 

+1

**Comment by ahus1** [comment]
@pedroigor - replying to your comment:

> For instance, checking both HTTPS and Hostname settings we should be able to print a warning in the logs.

Unfortunately in cloud environments this all depends on the proxy or if the browser is using a localhost or equivalent. All of this is undetectable by the server by looking at the CLI options only. The only real test is if the client provides the cookies, which might require one more round-trip.

**Comment by ahus1** [comment]
I pushed one more commit to set the same attributes on the `KC_RESTART` cookies. I think that makes sense for this one to have the same as the other authentication cookies, as it enhances security.

Once I did that, Keycloak shows once I submit a login form the following message (see screen shot at the end): 

> We are sorry...
Cookie not found. Please make sure cookies are enabled in your browser.

@jonkoops - assuming this is the `error.ftl` template, could we add to it something to it along the following lines: 

* JavaScript to detect if this page is a "secure environment" according to the browser, as IMHO it can't be reliably determined on the server side
* Have an explanation what is happening? The current text "Please make sure cookies are enabled in your browser." doesn't make the cut, so I'd like to have an additional text that shows conditionally on the error message code, and is localized in the `messages_en.properties`.

@jonkoops - if you could supply me with a snippet of JavaScript and a possible text to display in the scenario described above, I can have a look how to pull the text out of the localized message files (haven't done that before, but I suppose I'll figure it out).

image

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (jonkoops) [reply docs/documentation/release_notes/topics/24_0_0.adoc:9]: > Yeah, but only if we don't set the attribute, right?

To my knowledge also if we don't set the attribute, it will eventually not be possible to share cookies in a third-party context.

**Comment by pedroigor** [comment]
> I pushed one more commit to set the same attributes on the `KC_RESTART` cookies. I think that makes sense for this one to have the same as the other authentication cookies, as it enhances security.

This one in particular should be a first-party cookie all the time, I think. It is supposed to be used within the server domain boundaries and not shared across domains?

If browsers are going to default to `strict` then it should be better to rely on the default or force this one as `strict`.

(That is kinda why I think we should avoid introducing strict in this PR because there are other things to discuss)

**Comment by ahus1** [comment]
@jonkoops - pushed one more commit that uses the JavaScript you suggested to have a message like the following: 

> Cookie not found. Make sure cookies are enabled in your browser and the page is running in a secure context.
At the moment, this page is not running in a secure context. To run in a secure context it requires a URL starting with https://. In development environments, also URLs like http://127.0.0.1, http://localhost or any URL ending with .localhost like http://example.localhost would work, but this is not supported for production environments.

As it might run in an enterprise environment, it shouldn't require a link to the Internet. 

image

**Comment by ahus1** [comment]
> This one in particular should be a first-party cookie all the time, I think. It is supposed to be used within the server domain boundaries and not shared across domains?

Reading this, I'm now confused why the other cookies have `None` set. Still, won't question this as part of this PR. 

I've now set the `KC_RESTART` cookie to have `secure` set to `true`, which has the same effect and doesn't touch the samesite stuff in one more commit.

**Comment by pedroigor** [comment]
> Reading this, I'm now confused why the other cookies have `None` set. Still, won't question this as part of this PR.

Same here ...  However, an authentication session is used in brokering and there you have cross-origin requests where the user is redirected back to the server to continue with the flow.

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (mposolda) [inline docs/documentation/release_notes/topics/24_0_0.adoc:15]: Does this mean that some more update might need to be done in the documentation in relation to this? I can think of some update for the option `http-enabled` documented here https://www.keycloak.org/server/enabletls (as per my understanding `http-enabled` won't work correctly unless people are running on `localhost` or legacy cookies feature is enabled).

**Review by mposolda — COMMENTED** [review]
Added few minor comments to the docs part.

**Review by stianst — COMMENTED** [review]
We should probably add something to cleanup `*_LEGACY` cookies as otherwise these are left until deleted manually by the user, or they expire (which could be a long time if remember-me is enabled).

**Comment by stianst** [comment]
I wonder if we should add some documentation/guidance around cookies, as the migration guides now contain useful tips to those reading those, but not to new users.

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (jonkoops) [reply docs/documentation/release_notes/topics/24_0_0.adoc:15]: > Does this mean that some more update might need to be done in the documentation in relation to this?

Yes, the documentation will have to be reviewed and modified in this regard. Once the implementation settles down that will be the next step.

**Comment by stianst** [comment]
I'm not a fan of the following:

image

This is not really information aimed at the user, but rather to the admin/ops guy installing Keycloak. Would be better to just display a generic error page, and output an error to the log.

**Comment by jonkoops** [comment]
> I wonder if we should add some documentation/guidance around cookies, as the migration guides now contain useful tips to those reading those, but not to new users.

Yes, see https://github.com/keycloak/keycloak/pull/26161#discussion_r1453056793. We'll have to dedicate a section of the getting started guide to ensuring users run Keycloak in a secure context.

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (pedroigor) [reply docs/documentation/release_notes/topics/24_0_0.adoc:9]: Not really for applications running thin the same domain, right?

We have open issues about making it possible to configure our cookies and use `strict`. As I mentioned before, for now we should keep using `none` and have a follow-up to properly discuss the best/secure default settings. But I think it makes sense to make this configurable too.

**Review by stianst — CHANGES_REQUESTED** [review]
Summary of issues from me:

* We can not use `strict` (as that restricts access on redirects as well) - at this point this is only a note in the release notes, and any changes here should be done in a follow-up so straightforward thing is just remove this mention in the release notes for now
* Support for insecure contexts dropped - this could probably be explained a little bit better in the release notes, as it's not obvious that it's the legacy cookies that somewhat works around this issue currently
* Lack of cleanup of *_LEGACY cookies
* Error message. I don't link the specific hard-coded bits for this error message in the error form) and it also shows more information to the end-user than it should

**Comment by stianst** [comment]
After chatting with @jonkoops we have some slightly different ideas to try out first, in either case we'll not target removing legacy cookies for 24, but rather look at getting this into 25. Leaving the PR open here for now.

**Pull request closed without merging**
